### PR TITLE
Add --lines to follow

### PR DESCRIPTION
**Pull request by drewkett**
This adds a `--lines` option to the follow command. I also added a more efficient implementation of read_last_lines to `pueue-lib` (compared to the simple one in there now). I branched off `pueue-lib` off of `0.18.1` because it was the quickest way for me to test it since I wasn't sure if you've made any breaking changes in the work towards the new version. I'll make a separate PR for `pueue-lib` for visibility (Nukesor/pueue-lib#18). If you're interested in merging these PR's, let me know the preferred method of getting it incorporated with respect to which branch to use as the base and how to coordinate `pueue` and `pueue-lib` changes. I also see you're working on a new version so maybe this should just target that release? Let me know

### CODE REVIEW

**Comment by Nukesor** [comment]
I decided to pull the `pueue-lib` project back into the main repository!

Could you merge rebase onto the current `development` HEAD your changes in pueue-lib into this MR?
This will should make colaburation much easier :)

**Comment by drewkett** [comment]
That definitely simplifies this change. This should bow be up to date on the development branch. It compiles and tests fine but haven't actually tested the functionality again with the move to the development branch.

File: lib/src/log.rs
Comment (Nukesor) [inline lib/src/log.rs:137]: Pueue isn't used on 16bit platforms. At least I'm not aware of it.
Anyhow, u64 should still be supported u16 systems anyway. There are also u128 on 64-bit systems. The number will then just occupy four memory addresses instead of a single one.

I think it's best to switch this to `u64`. This will remove a lot of number casting and error handling.

File: lib/src/log.rs
Comment (Nukesor) [inline lib/src/log.rs:133]: Missing docstring.
This function contains a quite complex algorithm. This should be well documented.

File: lib/src/log.rs
Comment (Nukesor) [inline lib/src/log.rs:140]: Please don't use single-character variable names.
Especially if they're rather confusing. Please rename `p` to `size` or `seeked_bytes`.

File: lib/src/log.rs
Comment (Nukesor) [inline lib/src/log.rs:148]: Please add comments to runtime variables.
Or make the variable names more explicit, such as `cursor_position`.

File: lib/src/log.rs
Comment (Nukesor) [inline lib/src/log.rs:149]: Same here.
`text` could be everything. `last_lines` would be a much better fit.

File: lib/src/log.rs
Comment (Nukesor) [inline lib/src/log.rs:163]: This is very equal to the current behavior of https://doc.rust-lang.org/std/io/struct.BufReader.html.

I checked whether there already is a crate for this and it turns out there is!
https://docs.rs/rev_buf_reader/latest/rev_buf_reader/

The project is rather small, but it is very well documented and most importantly properly tested.

To be honest, I don't feel too comfortable to include such a finnicky algorithm, if there's a proper and tested implementation out there. Just like the official BufReader it can also read lines in reverse.

I made a small playground project:

Cargo.toml:
```
[package]
name = "rofl"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rev_buf_reader = "0.2"
```

main.rs: 
```
use rev_buf_reader;

use rev_buf_reader::RevBufReader;
use std::io::{self, BufRead};

fn main() {
    let data = "This\nis a long\ntest sentence";
    let inner = io::Cursor::new(&data);
    let reader = RevBufReader::new(inner);
    let mut lines = reader.lines();

    assert_eq!(lines.next().unwrap().unwrap(), "test sentence".to_string());
    assert_eq!(lines.next().unwrap().unwrap(), "is a long".to_string());
    assert_eq!(lines.next().unwrap().unwrap(), "This".to_string());
    assert!(lines.next().is_none());
}
```

File: client/display/follow.rs
Comment (Nukesor) [inline client/display/follow.rs:37]: A comment explaining what's happening in here would be nice.

File: lib/src/log.rs
Comment (Nukesor) [reply lib/src/log.rs:163]: This implementation will be a little less performant as we're no longer operating on the byte level, but also introduce UTF-8 handling. However, performance usually isn't an issue in these cases. A few `μs` on top won't hurt us.

We should also be able to safely use the `utf8_from_lossy` function. It replaces invalid characters with the Placeholder character, which would be done by the terminal lateron anyway.

**Review by Nukesor — CHANGES_REQUESTED** [review]
First of all thanks for your contributions (Including your other PR) :) Welcome to the project.

In general this looks fine, but it would be awesome if you could evaluate whether the rev_buf_reader crate serves this usecase.

This could potentiall us safe a lot of code and replace it with properly tested and well documented code :)

**Comment by Nukesor** [comment]
You can probably ignore most of the comments, if the `rev_buf_reader` works. I only noticed the similarity to the BufReader halfway through the review.

**Comment by Nukesor** [comment]
I went ahead and double-checked the `rev_buf_reader` crate.
In general it looks good to me beside a few little issues.

I created a few PRs (https://github.com/andre-vm/rev_buf_reader/pull/1), let's see how active the maintainer is :)
We could always offer a co-maintainership.

**Comment by Nukesor** [comment]
I just noticed, that this is also missing the logic for streaming the content from the daemon to the client.
It looks like you only implemented the client side local reading.

There's a separate file in `daemon/network/follow.rs`, that's used to stream log output from a remote daemon to clients.

File: lib/src/log.rs
Comment (Nukesor) [inline lib/src/log.rs:201]: After reading through my old docs and the old implementation, I'm not 100% sure if your current implementation changes anything at all.
`rev_lines` also read the file from the back in a BufReader style, but with a small convenience layer on top.

The part about imperformant code mentioned in the docstring isn't about the bufreader implementation, but rather about reading all lines into memory, reversing them and only then streaming them to disk.

This gets troublesome, if users try to read a few thousand lines with possibly gigabytes of data from the logs. This already happened and lead to RAM exhaustion.

A proper performant solution would be to read through the file until X+1 `\n` are found.
The character after the `\n` is then the start of our log output and we could simply pass the handle to a io::copy call with stdout as the sink.
`\n\r` should obviously also be handled

This function would then be changed to something like this:
```
/// Seek to the beginning of a line, which is located `lines` newlines (`\n`) from the end of the file.
/// Jumps to the top of the files, if it doesn't contain enough files.
/// Returns the number of lines that were found.
pub fn seek_to_line_rev(file: &mut File, lines: usize) -> Result<usize> {}
```

Comment (drewkett) [reply lib/src/log.rs:201]: FWIW, I believe my implementation does in fact read in 4k chunks going backwards. I agree generally it could use more documentation, but I sort of just submitted this as first pass to see if you had interest in using it. That said, I do agree that using that crate probably makes more sense than the implementation I came up with.

Comment (Nukesor) [reply lib/src/log.rs:201]: It does so, but this is a usual behavior of a BufReader.
The problem isn't the chunked reading, but rather the fact that it collects them into memory before streaming them to stdout.

The performant way would be to seek to the desired position in the file and stream the remaining contents directly to stdout.

It's a "read into memory and copy to stdout" vs. "stream directly to stdout". Reading into memory can be critical with the amount of data that possibly accrues

File: lib/src/log.rs
Comment (drewkett) [reply lib/src/log.rs:137]: If I used `u64` though, I believe I would need to do `try_into` or something similar on 32-bit platforms to be safe for the cast to `usize` (obviously the chunk size won't be that big). Thats what I was trying to avoid. Though its possible all my casting is less readable than using `try_into` . So even if we drop 16-bit, to keep my code as is, it could only be `u32`.

File: lib/src/log.rs
Comment (drewkett) [reply lib/src/log.rs:133]: Agreed. I'll write one up if we end up keeping this function instead of using the other crate

File: lib/src/log.rs
Comment (drewkett) [reply lib/src/log.rs:199]: I see what you're saying now. Yes that is true that I didn't try to address the scenario where a large number of lines are requested. It doesn't strike me as a likely use case, but understood that ideally that would be addressed as well. One thing that stopped me from even looking at that was that I wasn't sure of is whether there is currently a mechanism to stream data between client/server.  Since its message based, I would think there would need to be a mechanism for sending data in chunks from server to client which didn't jump out at me in the code but I also may have just missed it.

That said using that other crate with everything else the same, should at least prevent the extra copy of data in `read_last_lines` since it looked like it stores everything contiguously as opposed to my `VecDeque` based implementation.

**Comment by drewkett** [comment]
Thanks for all the comments @Nukesor. I realize I should have cleaned this up a bit more before just dumping it as a PR. I also  appreciate having you go through my code and point out some of the stylistic things that I do that are not exactly best practices.

File: lib/src/log.rs
Comment (drewkett) [reply lib/src/log.rs:199]: Ignore the last comment regarding streaming, of course there is a mechanism to do that. I 

> I went ahead and double-checked the `rev_buf_reader` crate. In general it looks good to me beside a few small issues.
> 
> I created a few PRs ([andre-vm/rev_buf_reader#1](https://github.com/andre-vm/rev_buf_reader/pull/1)), let's wait and see if the maintainer is still active :) We could always offer a co-maintainership.

Just glanced at this PR and it seems like its maintained based on the responses, so I'm going to go ahead and switch to it. I'll do a quick review of that crate as well just to have another set of eyes on.

Comment (drewkett) [reply lib/src/log.rs:199]: Ignore the last comment regarding streaming, of course there is a mechanism to do that. I 

> I went ahead and double-checked the `rev_buf_reader` crate. In general it looks good to me beside a few small issues.
> 
> I created a few PRs ([andre-vm/rev_buf_reader#1](https://github.com/andre-vm/rev_buf_reader/pull/1)), let's wait and see if the maintainer is still active :) We could always offer a co-maintainership.

Just glanced at this PR and it seems like its maintained based on the responses, so I'm going to go ahead and switch to it. I'll do a quick review of that crate as well just to have another set of eyes on.

**Comment by drewkett** [comment]
In switching to `rev_buf_reader`, how much do you care about performance? The most straightforward implementation as you showed above would allocate a new `String` for every line as it iterates back through the number of lines. An alternative that should be more performant would be to create a function to seek back in the file the requested number of lines and then stream the data as you suggested in another comment. But that implementation would be more complex.

File: lib/src/log.rs
Comment (Nukesor) [reply lib/src/log.rs:137]: Ah I see. You're right about that. Casting to usize wouldn't be safe in that case.

**Comment by Nukesor** [comment]
This is quite a fun problem :D

I tinkered around a bit and came up with this solution. I already tried it and it seemed to work like a charm.

```
/// Seek the current file handle to the beginning of the line that's located `amount` newlines
/// from the back of the file.
pub fn seek_to_last_lines(file: &mut File, amount: usize) -> Result<(), Error> {
    let mut reader = RevBufReader::new(file);
    // The position from which the RevBufReader starts reading.
    // The file size might change while we're reading the file. Hence we have to save it now.
    let start_position = reader.get_mut().seek(SeekFrom::Current(0))?;
    let start_position: i64 = start_position.try_into().map_err(|_| {
        Error::Generic("Failed to convert start cursor position to i64".to_string())
    })?;
    println!("Start position: {}", start_position);

    let mut total_read_bytes: i64 = 0;
    let mut found_lines = 0;

    // Read in 4KB chunks until there's either nothing left or we found `amount` newline characters.
    loop {
        let mut buffer = vec![0; 4096];
        let read_bytes = reader.read(&mut buffer)?;

        // Return if there's nothing left to read.
        // We hit the start of the file and read fewer lines then specified.
        if read_bytes == 0 {
            break;
        }

        // Check each byte for a newline.
        // Register all newlines we find.
        // Even though the RevBufReader reads from behind, the bytes are still in the correct
        // order. Since we want to scan from the back, we have to reverse the buffer
        for byte in buffer[0..read_bytes].iter().rev() {
            total_read_bytes += 1;
            println!("char: '{:?}'", *byte as char);
            if *byte != b'\n' {
                continue;
            }

            // We found a newline.
            println!(
                "Found newline at byte: {}",
                start_position - total_read_bytes
            );
            found_lines += 1;

            // We found the requested amount of lines.
            if found_lines == amount + 1 {
                println!("Found {} lines", amount);
                // The RevBufReader most likely already went past this point.
                // That's why we have to set the cursor to the position of the last newline.
                println!("Total read byte: {}", total_read_bytes);

                // Calculate the distance from the start to the desired location.
                let distance_to_start = start_position - total_read_bytes + 1;
                // Cast it to u64. If it somehow became negative, just seek to the start of the
                // file.
                let distance_to_start: u64 = distance_to_start.try_into().unwrap_or(0);
                println!("Distance: {}", distance_to_start);

                // We can safely unwrap `start_position`, as we previously casted it from an u64.
                if distance_to_start < start_position.try_into().unwrap() {
                    // Seek to the position.
                    let inner = reader.into_inner();
                    let current_position = inner.seek(SeekFrom::Start(distance_to_start))?;
                    println!("Final position: {}", current_position);
                }

                return Ok(());
            }
        }
    }

    Ok(())
}
```

**Comment by Nukesor** [comment]
I went ahead and implemented the new seeking logic in a separate PR.
This PR also fixes some other log related issues and adds proper tests for this behavior.
https://github.com/Nukesor/pueue/pull/278

Adding your feature with these changes should now be trivial :)

**Comment by drewkett** [comment]
Definitely makes this change simpler 👍. I cleaned up the commit history and stripped it down to just this change. Let me know if I missed anything

**Comment by Nukesor** [comment]
Perfect, this looks good to me :)

Thanks for your contribution and sorry for the constant back and forth ;D

**Review by Nukesor — APPROVED** [review]

**Pull request merged**
